field/composite_test.go

```diff
@@ -952,6 +1042,309 @@ func TestCompositePackingWithTags(t *testing.T) {
+	})
+
+	t.Run("Unpack returns an error on data having subfield ID not in spec with default bitmap", func(t *testing.T) {
+		data := &CompositeTestData{}
+
+		composite := NewComposite(compositeTestSpecWithDefaultBitmap)
+		err := composite.SetData(data)
+		require.NoError(t, err)
+
+		// Index 2-3 = 70 indicates the presence of field 4. This field is not defined on spec.
+		read, err := composite.Unpack([]byte("32702000000000000002AB0212060102YZ"))
+		require.Equal(t, 0, read)
+		require.EqualError(t, err, "failed to unpack subfield 4: no specification found")
+	})
+
+	t.Run("Unpack correctly deserialises out of order composite subfields to the data struct with default bitmap", func(t *testing.T) {
+		composite := NewComposite(compositeTestSpecWithDefaultBitmap)
+
+		read, err := composite.Unpack([]byte("32602000000000000002AB0212060102YZ"))
+
+		require.NoError(t, err)
+		require.Equal(t, 34, read)
+
+		data := &CompositeTestData{}
+		require.NoError(t, composite.Unmarshal(data))
+
+		require.Equal(t, "AB", data.F2.Value())
+		require.Equal(t, 12, data.F3.Value())
+		require.Equal(t, "YZ", data.F11.F1.Value())
+	})
+
+	t.Run("Unpack correctly deserialises partial subfields to the data struct with default bitmap", func(t *testing.T) {
+		composite := NewComposite(compositeTestSpecWithDefaultBitmap)
+
+		read, err := composite.Unpack([]byte("24600000000000000002AB0212"))
+
+		require.NoError(t, err)
+		require.Equal(t, 26, read)
+
+		data := &CompositeTestData{}
+		require.NoError(t, composite.Unmarshal(data))
+
+		require.Equal(t, "AB", data.F2.Value())
+		require.Equal(t, 12, data.F3.Value())
+		require.Nil(t, data.F11)
+	})
+
+	t.Run("Unpack correctly ignores excess bytes in excess of the length described by the prefix with default bitmap", func(t *testing.T) {
+		composite := NewComposite(compositeTestSpecWithDefaultBitmap)
+
+		// "60102YZ" falls outside of the bounds of the 24 byte limit imposed
+		// by the prefix. Therefore, F11 must be nil.
+		read, err := composite.Unpack([]byte("24600000000000000002AB0212060102YZ"))
+
+		require.NoError(t, err)
+		require.Equal(t, 26, read)
+
+		data := &CompositeTestData{}
+		require.NoError(t, composite.Unmarshal(data))
+
+		require.Equal(t, "AB", data.F2.Value())
+		require.Equal(t, 12, data.F3.Value())
+		require.Nil(t, data.F11)
+	})
+
+	t.Run("Unpack returns an error on failure of subfield to unpack bytes with sized bitmap on 3 bytes", func(t *testing.T) {
+		data := &CompositeTestData{}
+
+		composite := NewComposite(compositeTestSpecWithSizedBitmap)
+		err := composite.SetData(data)
+		require.NoError(t, err)
+
+		// F2 fails to unpack - it requires length to be defined instead of AB.
+		read, err := composite.Unpack([]byte("16602000AB060102YZ"))
+		require.Equal(t, 0, read)
+		require.Error(t, err)
+		require.EqualError(t, err, "failed to unpack subfield 2 (String Field): failed to decode length: strconv.Atoi: parsing \"AB\": invalid syntax")
```

Comment:
I think that in this case we potentially can leak sensitive data and I suggest using `SafeError` here. You can find examples of how we use it in the iso858 package.

Reply:
https://github.com/moov-io/iso8583/blob/49dc79f0a98c23673bc1baaabb3fd1d986175ece/encoding/ascii.go#L20

but I've checked that in your case it's not needed. We convert index/tag which is expected to be number and is in the spec only, not in the message data.